main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c

```diff
@@ -0,0 +1,1192 @@
+            payload.selectedSubgame.x = ttt->game.selectedSubgame.x;
+            payload.selectedSubgame.y = ttt->game.selectedSubgame.y;
+
+            do
+            {
+                if (payload.cursor.x < ttt->game.cpu.destCell.x && payload.cursor.x < 2)
+                {
+                    payload.cursor.x++;
+                }
+                else if (payload.cursor.y < ttt->game.cpu.destCell.y && payload.cursor.y < 2)
+                {
+                    payload.cursor.y++;
+                }
+                else if (payload.cursor.x > ttt->game.cpu.destCell.x && payload.cursor.x > 0)
+                {
+                    payload.cursor.x--;
+                }
+                else if (payload.cursor.y > ttt->game.cpu.destCell.y && payload.cursor.y > 0)
+                {
+                    payload.cursor.y--;
+                }
+                else
+                {
+                    if (!tttCursorIsValid(ttt, &ttt->game.cursor))
+                    {
+                        TCPU_LOG("??????????????? cursor is not valid????????\n");
+                    }
+                    // The cursor is in the right place, select it!
+                    tttMsgPlaceMarker_t placePayload;
+                    placePayload.type              = MSG_PLACE_MARKER;
+                    placePayload.selectedSubgame.x = ttt->game.selectedSubgame.x;
+                    placePayload.selectedSubgame.y = ttt->game.selectedSubgame.y;
+                    placePayload.selectedCell.x    = ttt->game.cpu.destCell.x;
+                    placePayload.selectedCell.y    = ttt->game.cpu.destCell.y;
+
+                    ttt->game.cpu.state = TCPU_INACTIVE;
+
+                    tttReceivePlacedMarker(ttt, &placePayload);
+
+                    return;
+                }
+            } while (!tttCursorIsValid(ttt, &payload.cursor));
+
+            tttReceiveCursor(ttt, &payload);
+        }
+        else
+        {
+            TCPU_LOG("Something else???");
+        }
+    }
+}
+
+static bool selectSubgame_easy(ultimateTTT_t* ttt, int* x, int* y)
+{
+    int availableSubgames[9] = {0};
+    int availableCount       = 0;
+
+    for (int idx = 0; idx < 9; idx++)
+    {
+        if (TTT_NONE == ttt->game.subgames[idx % 3][idx / 3].winner)
+        {
+            TCPU_LOG("Sub-game %d, %d is a valid target", idx % 3, idx / 3);
+            availableSubgames[availableCount++] = idx;
+        }
+    }
+
+    // Pick a random subgame
+    int subgameIndex = availableSubgames[esp_random() % availableCount];
+
+    *x = subgameIndex % 3;
+    *y = subgameIndex / 3;
+    return true;
+}
+
+static bool selectSubgame_medium(ultimateTTT_t* ttt, int* x, int* y)
+{
+    // Still just random
+    return selectSubgame_easy(ttt, x, y);
+}
+
+static bool selectSubgame_hard(ultimateTTT_t* ttt, int* x, int* y)
+{
+    // Eventually this will become the medium difficulty once I get the actual difficult AI working
```

Comment:
Is this still pending?

Reply:
Hmm, well I still do want to do this but I'm not sure if I can commit to it until I get some time to actually work on it... I think the current "Hard" AI is at least a worthy opponent in that it doesn't make super obvious errors and it plays well enough to make it interesting. But it doesn't actually feel that hard to me, so I still do want to try and write a more advanced AI that actually thinks a few moves ahead, and then either have parameters to tune its difficulty for the other levels, or shift everything down a level and drop the pure-random difficulty.

Reply:
Ah, gotcha. I'll consider that a stretch goal, but not a top priority.